java-checks/src/main/java/org/sonar/java/checks/AnonymousClassShouldBeLambdaCheck.java

```diff
@@ -69,21 +74,53 @@ public void visitNewClass(NewClassTree tree) {
     if (classBody != null) {
       TypeTree identifier = tree.identifier();
-      if (!useThisIdentifier(classBody) && !enumConstants.contains(identifier) && isSAM(classBody)) {
+      if (!useThisIdentifier(classBody) && !enumConstants.contains(identifier) && isSAM(classBody, identifier)) {
         context.reportIssue(this, identifier, "Make this anonymous inner class a lambda" + context.getJavaVersion().java8CompatibilityMessage());
       }
     }
   }
 
-  private static boolean isSAM(ClassTree classBody) {
-    if(hasOnlyOneMethod(classBody.members()) && classBody.symbol().isTypeSymbol()) {
+  private static boolean isSAM(ClassTree classBody, TypeTree identifier) {
+    List<Tree> members = classBody.members();
+    if (hasOnlyOneMethod(members) && classBody.symbol().isTypeSymbol()) {
       // Verify class body is a subtype of an interface
       JavaSymbol.TypeJavaSymbol symbol = (JavaSymbol.TypeJavaSymbol) classBody.symbol();
-      return symbol.getInterfaces().size() == 1 && symbol.getSuperclass().is("java.lang.Object");
+      MethodTree method = (MethodTree) members.get(0);
+      JavaSymbol.TypeJavaSymbol interfaceOwner = (JavaSymbol.TypeJavaSymbol) identifier.symbolType().symbol();
+      return symbol.getInterfaces().size() == 1 &&
+        symbol.getSuperclass().is("java.lang.Object") &&
+        notMultipleDefaultMethodsWithSameSignature((JavaSymbol.MethodJavaSymbol) method.symbol(), interfaceOwner);
     }
     return false;
   }
 
+  private static boolean notMultipleDefaultMethodsWithSameSignature(JavaSymbol.MethodJavaSymbol methodSymbol, JavaSymbol.TypeJavaSymbol interfaceOwner) {
+    Set<JavaSymbol.MethodJavaSymbol> methods = new HashSet<>();
+    List<TypeJavaSymbol> allInterfacesOfHierarchy = interfaceOwner.superTypes().stream().filter(type -> !type.is("java.lang.Object"))
+      .map(ClassJavaType::getSymbol).collect(Collectors.toList());
+    for (JavaSymbol.TypeJavaSymbol interfaceOfHierarchy : allInterfacesOfHierarchy) {
+      methods.addAll(interfaceOfHierarchy.memberSymbols().stream().filter(Symbol::isMethodSymbol).map(method -> ((JavaSymbol.MethodJavaSymbol) method))
+        .filter(method -> !method.name().equals(methodSymbol.name()))
+        .filter(method -> sameParameters(method, methodSymbol))
+        .collect(Collectors.toList()));
+    }
+    methods.addAll(
+      interfaceOwner.memberSymbols().stream().filter(Symbol::isMethodSymbol).map(method -> ((JavaSymbol.MethodJavaSymbol) method))
+        .filter(method -> sameParameters(method, methodSymbol))
+        .collect(Collectors.toList()));
+    return methodsDefaultAndNotStatic(methods);
+  }
+
+  private static boolean methodsDefaultAndNotStatic(Set<JavaSymbol.MethodJavaSymbol> methods) {
+    List<JavaSymbol.MethodJavaSymbol> methodsRemained = methods.stream().filter(method -> !method.isStatic() && method.isDefault())
+      .collect(Collectors.toList());
+    return methodsRemained.isEmpty() || methodsRemained.size() <= 1;
```

Comment:
`methodsRemained.size() <= 1` is enough, because equivalent to `empty && size==1`, no need the `methodRemained.isEmpty()` then.

And consequently, you can replace everything by:
`return methods.stream().filter(JavaSymbol.MethodJavaSymbol::isDefault).count() <= 1;`

And you can then inline the method and remove this one.

Reply:
Now, based on compilable results from test files (please update), it seems that it does not work.